edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/PodExtensions.cs

```diff
@@ -43,35 +44,25 @@ static Option<V1ContainerStatus> GetContainerByName(string name, V1Pod pod)
             return Option.Maybe(status);
         }
 
-        static ReportedModuleStatus ConvertPodStatusToModuleStatus(V1PodStatus podStatus)
+        static ReportedModuleStatus ConvertPodStatusToModuleStatus(Option<V1PodStatus> podStatus)
         {
-            ModuleStatus moduleStatus = ModuleStatus.Failed;
-            string description = "Invalid pod status";
-
-            if (podStatus?.Phase != null)
-            {
-                switch (podStatus.Phase)
-                {
-                    case "Running":
-                        moduleStatus = ModuleStatus.Running;
-                        description = $"Started at {podStatus.StartTime}";
-                        break;
-                    case "Failed":
-                    case "Pending":
-                    case "Unknown":
-                        moduleStatus = ModuleStatus.Failed;
-                        description = podStatus.Reason;
-                        break;
-                    case "Succeeded":
-                        moduleStatus = ModuleStatus.Stopped;
-                        description = podStatus.Reason;
-                        break;
-                    default:
-                        throw new InvalidOperationException("Invalid pod status");
-                }
-            }
-
-            return new ReportedModuleStatus(moduleStatus, description);
+            return podStatus.Map(
+                status =>
+                {
+                    switch (status.Phase)
+                    {
+                        case "Running":
+                            return new ReportedModuleStatus(ModuleStatus.Running, $"Started at {status.StartTime}");
+                        case "Failed":
+                        case "Pending":
+                        case "Unknown":
+                            return new ReportedModuleStatus(ModuleStatus.Failed, status.Reason);
+                        case "Succeeded":
+                            return new ReportedModuleStatus(ModuleStatus.Stopped, status.Reason);
+                        default:
+                            throw new InvalidOperationException($"Invalid pod status {status.Phase}");
+                    }
+                }).GetOrElse(() => throw new InvalidParameterException("No pod status"));
```

Comment:
Not sure that throwing an exception when an option is none is a very good idea. Basically, an exception describes impossible to handle a situation, right? So if you think podStatus can be None in some cases, we should handle this case in a way that it doesn't lead to a crash.

Reply:
talked offline, changed to returning a failed status with a custom description.

---
In reply to: [341776800](https://github.com/Azure/iotedge/pull/1938#discussion_r341776800) [](ancestors = 341776800)